Code/GraphMol/Sgroup.h

```diff
@@ -0,0 +1,215 @@
+
+namespace RDKit {
+class ROMol;
+class Bond;
+class Atom;
+
+//! used to indicate errors from incorrect sgroup access
+class RDKIT_GRAPHMOL_EXPORT SGroupException : public std::runtime_error {
+ public:
+  //! construct with an error message
+  SGroupException(const char *msg) : std::runtime_error(msg){};
+  //! construct with an error message
+  SGroupException(const std::string &msg) : std::runtime_error(msg){};
+};
+
+class RDKIT_GRAPHMOL_EXPORT SGroup {
+ public:
+  enum class BondType {
+    XBOND,  // External
+    CBOND,  // Internal
+  };
+
+  typedef std::array<RDGeom::Point3D, 3> Bracket;
+
+  struct AttachPoint {
+    Atom *aAtom;
+    Atom *lvAtom;
+    std::string id;
+  };
+
+  struct CState {
+    Bond *bond;
+    boost::shared_ptr<RDGeom::Point3D> vector;
+  };
+
+  typedef std::vector<Atom *> ATOM_PTR_VECT;
+  typedef ATOM_PTR_VECT::iterator ATOM_PTR_VECT_I;
+  typedef ATOM_PTR_VECT::const_iterator ATOM_PTR_VECT_CI;
+  typedef std::vector<Bond *> BOND_PTR_VECT;
+  typedef BOND_PTR_VECT::iterator BOND_PTR_VECT_I;
+  typedef BOND_PTR_VECT::const_iterator BOND_PTR_VECT_CI;
+
+  //! No default constructor
+  SGroup() = delete;
+
+  //! Constructor. Ownsership is only set on this side of the
+  //! relationship: mol->addSGroup(sgroup) still needs to be
+  //! called to get ownership on the other side.
+  SGroup(ROMol *owning_mol, const std::string &_type);
+
+  //! Copy Constructor: initialize from a second sgroup.
+  SGroup(const SGroup &other) = default;
+
+  //! Destructor
+  ~SGroup(){};
+
+  //! Get the molecule that owns this conformation
+  ROMol &getOwningMol() const { return *dp_mol; }
+
+  //! get the ID of this sgroup
+  inline unsigned int getId() const { return d_id; }
+
+  //! get the COMPNO of this sgroup
+  inline unsigned int getCompNo() const { return d_compno; }
+
+  //! get the index of this sgroup in dp_mol's sgroups vector
+  //! (do not mistake this by the ID!)
+  unsigned int getIndexInMol() const;
+
+  //! get the type of the SGroup
+  inline const std::string &getType() const { return d_type; }
+
+  //! check if SGroup has the given property set
+  bool hasProp(const std::string &prop) const;
+
+  //! get string properties
+  const std::string &getProp(const std::string &prop) const;
+
+  //! get parent SGroup
+  inline SGroup *getParent() const { return d_parent; }
+
+  //! set the ID of this sgroup
+  void setId(unsigned int id);
+
+  //! set the COMPNO of this sgroup
+  void setCompNo(unsigned int compno) { d_compno = compno; }
+
+  //! set the type of the SGroup
+  inline void setType(const std::string &type) { d_type = type; }
+
+  //! set string properties
+  inline void setProp(const std::string &prop, const std::string &value) {
+    d_prop[prop] = value;
+  };
+
+  //! set the parent SGroup
+  inline void setParent(SGroup *parent) { d_parent = parent; }
+
+  /* Atom and Bond methods */
+  void addAtomWithIdx(unsigned int idx);
+  void addPAtomWithIdx(unsigned int idx);
```

Comment:
I like explicit
```
addParentAtomWithIdx
```

Reply:
That's reasonable. I guess I had the SD spec codes in mind when I named these functions.